lib/client.js

```diff
@@ -0,0 +1,156 @@
+      var buf = muxBuffer(msgcode.stdin, new Buffer(data));
+      this.queue(buf);
+    });
+
+    client.bufin = through();
+    client.stdin.pipe(client.bufin);
+
+    client.bufin.pause();
+    client.finalin = through((data) => {
+      client.socket.send(data, {binary: true});
+    });
+    client.bufin.pipe(client.finalin);
+
+    client.stdout = through();
+    client.stderr = through();
+
+    //When the connection is paused, this listener is on
+    function paused (messageEvent) {
+      var message = new Uint8Array(messageEvent.data);
+      switch (message[0]) {
+        //resumes the client, flushing stdin
+        case msgcode.resume:
+          client.state = msgcode.resume;
+          client.bufin.resume();
+          client.emit('resumed');
+          client.socket.onmessage = resumed;
+          break;
+
+        case msgcode.stdout:
+          client.stdout.write(message.slice(1));
+          // debug('client stdout needs drain: ',!tf);
+          break;
+
+        case msgcode.stderr:
+          client.stderr.write(message.slice(1));
+          // debug('client stderr needs drain: ',!tf);
+          break;
+
+        //first byte contains exit code
+        case msgcode.stopped:
+          client.emit('exit', message.readInt8(1));
+          client.stdout.end();
+          client.stderr.end();
+          break;
+
+        default:
+          debug('unknown msg code %s', message[0]);
+          throw new Error('unknown msg code');
+      }
+    }
+
+    //When the connection is resumed, this listener is on
+    function resumed (messageEvent) {
+      var message = new Buffer(new Uint8Array(messageEvent.data));
+      debug(message.length);
+      // the first byte is the message code
+      switch (message[0]) {
+        //pauses the client, causing stdin to buffer
+        case msgcode.pause:
+          client.state = msgcode.pause;
+          client.bufin.pause();
+          client.emit('paused');
+          client.socket.onmessage = paused;
+          break;
+
+        case msgcode.stdout:
+          client.stdout.write(message.slice(1));
+          // debug('client stdout needs drain: ',!tf);
+          break;
+
+        case msgcode.stderr:
+          client.stderr.write(message.slice(1));
+          // debug('client stderr needs drain: ',!tf);
+          break;
+
+        //first byte contains exit code
+        //maybe i want to do more here?
+        case msgcode.stopped:
+          client.emit('exit', message.readInt8(1));
+          break;
+
+        default:
+          debug('unknown msg code %s', message[0]);
+          throw new Error('unknown msg code');
+      }
+    }
+
+    //Starts out paused so that input isn't sent until server is ready
+    client.socket.onmessage = paused;
+    return client;
+  }
+
+  //pauses input coming in from server, useful if you're running out of memory on local (ie probably never)
+  pause () {
+    this.bufin.write(new Buffer([msgcode.pause]), {binary: true});
```

Comment:
So since we never change the buffer we can reuse a buffer from previously... and just make a single global `const` buffer...
```js
const PAUSE_MESSAGE = new Buffer([msgcode.pause]);

class DockerExecWebsocketClient extends events.EventEmitter {
  ....
  pause() {
    this.bufin.write(PAUSE_MESSAGE, {binary: true});
  }
}
```

This is an insignificant optimization... Just nice to know :)